### DISCUSSION TITLE
VenusOS NodeRed direct import over the ttyUSB0 Adapter?

### DISCUSSION THREAD

**Discussion by moench13** [syssi/esphome-pace-bms#24] [General]
Hello,
it is possible to the the R and RW input to the pace-bms driektliy over VenosOS/NodeRed with the USB Dongle RS485 or RS232 plugged in directly on the Cerbo or Raspberry PI? So you get the Values over the NodeRed palette "serialport or modbus"?
greets

**Comment by syssi** [comment]
Yes this should be possible but it's out of scope of this project.

**Comment by moench13** [comment]
mybe you can to a project like this :)

i just plugged the usb dongle to the raspberry... i tried this... is my thinking right...  i also can try with a waveshare modbus to TCP adapter... are this Registers Holding Reg?

<img width="943" alt="Screenshot 2023-12-19 at 07 39 04" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/12cb88ad-a782-4eda-a65b-8d08aeff282e">

<img width="648" alt="Screenshot 2023-12-19 at 07 39 09" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/b953ee18-5077-4625-b571-07354e428b27">

<img width="635" alt="Screenshot 2023-12-19 at 07 39 49" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/b93c2c0c-4821-4f8e-97dd-45f32406b73d">

**Reply by syssi** [reply]
Yes, your settings are looking good in general.

1. Baudrate 9600
2. Protocol Modbus RTU
3. Register address 1
    https://github.com/syssi/esphome-pace-bms/blob/main/esp8266-example.yaml#L91-L103
4. Quantity (Length) 1
     https://github.com/syssi/esphome-pace-bms/blob/main/esp8266-example.yaml#L97C17-L97C23
     https://esphome.io/components/number/modbus_controller.html

I assume it's important to talk to your BMS using RS485. RS232 won't work probably.

**Comment by moench13** [comment]
looking good... poll rate is at 1 sec. but it is slow and timeouts... BAT1 is Unit1 BAT2 is Unit2 - No readings from BAT2 same Protocol - Gobel Pack1 is Master and is connected at RS485A with the USB to the VenusOS NodeRed, Pack 1 is from RS485B to Pack2 RS485C! Can i use the Yaml in Node Red to Import all Registers at once? Any Idea for the slow polling rate? 


![Screenshot 2023-12-19 at 19 37 24](https://github.com/syssi/esphome-pace-bms/assets/140324069/d3ac6b36-cf83-4ac4-aeb5-cd89593fccfc)
![Screenshot 2023-12-19 at 22 32 55](https://github.com/syssi/esphome-pace-bms/assets/140324069/d0ac2c9f-b51c-46e8-a038-001c08718564)
<img width="933" alt="Screenshot 2023-12-19 at 22 33 31" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/ed321f88-1d59-405f-90a1-4f44f2e42275">
<img width="855" alt="Screenshot 2023-12-19 at 22 36 06" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/f9481d22-ced1-444a-bbd4-9ad05ff26416">

**Reply by syssi** [reply]
You could request 36 registers at once if you set address 0 and length 36. In this case you will receive one big response per second which needs to be decoded later on using NodeRed.

**Comment by moench13** [comment]
First of all thanks for your support… today I tried it with a Waveshare Modbus to TCP adapter… here the polling is fine and fast! With the USB over Venus OS I will try the 36reg!

One more question… I’m connected now with the first master BMS over RS485A… second battery I need also Connect over RS485A or should it be possible do use the parallel communication ports RS485B,C… CAN is also Connected to VenusOS

**Reply by syssi** [reply]
What do you mean by RS485A, RS485B and RS485C and how to you declare the first BMS as master?

**Reply by moench13** [reply]
do i need a parallel connection do each RS485A port with the Waveshare Adapter so i can read the second battery?

![IMG_2266](https://github.com/syssi/esphome-pace-bms/assets/140324069/8469593f-ebcb-4fdc-8f01-7707bbc0fedd)

**Reply by syssi** [reply]
Could you check the manual of your battery how to chain multiple batteries?

**Comment by moench13** [comment]
looking good so fare with the Waveshare Adapter… every Pack needs his own RS485 connection!

![IMG_2267](https://github.com/syssi/esphome-pace-bms/assets/140324069/ef22ade8-4485-4a4f-93aa-57ce28063828)
![IMG_2268](https://github.com/syssi/esphome-pace-bms/assets/140324069/da403246-4582-4e82-96f0-94caac13857d)
<img width="1736" alt="Screenshot 2023-12-20 at 21 31 24" src="https://github.com/syssi/esphome-pace-bms/assets/140324069/76b98f60-4f07-4202-aeb5-f22ae5618095">

**Reply by syssi** [reply]
Did you connect both RS485 ports (for monitoring) to the same bus / Waveshare Adapter or do you use one adapter per BMS now?

**Reply by moench13** [reply]
Yes 1 Waveshare Modbus to TCP Adapter and then 2 Connections parallel from the Modbus Adapter to the RS485A Port on the 2 Gobel Packs! And the CAN is Still connected to the Cerbo/Raspberry!

**Reply by syssi** [reply]
And the second responds using device address 2 now? Good job!

**Reply by moench13** [reply]
Yes… the polling over the Modbus TCP adapter is working perfect… can read and write all settings but the most important settings i can not finde! Charge Voltage limit and so on… They send me this link, I’m not an expert on Modbus 🙈

Next I will check the RS485 to USB directly over the VenusOS… that would be the best solution because no external hardware/power and so on is necessary! Normally it should work because the RS485 Power meter is also polling in short sequences!

 https://github.com/fancyui/Gobel-Power-RN-BMS-RS485-ModBus

![IMG_2279](https://github.com/syssi/esphome-pace-bms/assets/140324069/59a75d1f-3dd0-43f3-8927-d2c9a4f8394a)

**Reply by syssi** [reply]
The `RfnBMS Tools` application uses a RS232 connection to read/write the mentioned settings (at the screenshot)?

**Reply by syssi** [reply]
[This protocol description](https://github.com/fancyui/Gobel-Power-RN-BMS-RS485-ModBus) isn't Modbus. It's about a different protocol known from Seplos devices / Pylontech batteries. Your BMS is probably able to speak this protocol too if you change the RS485 protocol setting.

Unfortunately we cannot extract any important information from this description. It doesn't provide any details about Modbus register addresses or something like that.

**Reply by moench13** [reply]
that's the pace RS232 protocol - but here i also see no value for this... the Chg cut-off volt = CVL in the Victron System... it would be nice to set this value remotely... so you can start every 2-3 month a top balancing with a higher CVL and limit the Charge Voltage in the DVCC.

In the Rfn BMS Tool - I can set the Values for every Battery Pack... only connected at the first Master BMS on RS485A port... with the Modbus over NodeRed I need for each Pack a parallel connection to the Waveshare Adapter that's not the best solution! I'm thinking about making a RJ45 Box where i connect all Packs Parallel together an then a Uplink to the Waveshare or USB dongle.

[72651=11625-PACE-RS232commuciation-protocal（PACE-RS232-TY16S）-20180705.pdf](https://github.com/syssi/esphome-pace-bms/files/13752891/72651.11625-PACE-RS232commuciation-protocal.PACE-RS232-TY16S.-20180705.pdf)
